imagepaste/clipboard/darwin/darwin.py

```diff
@@ -51,8 +52,8 @@ def push(cls, save_directory: str) -> DarwinClipboard:
         commands = [
             '((clipboard info) as string does not contain "«class furl»") as string'
         ]
-        process = Process(cls.get_osascript_args(commands))
-        if process.stdout == "true":
+        process = Process.execute(cls.get_osascript_args(commands))
+        if process.stdout[0] == "true":
```

Comment:
Can we do this, it will save 2 lines:
```suggestion
        commands = ['((clipboard info) as string contain "«class furl»") as string']
        process = Process(cls.get_osascript_args(commands))
        if process.stdout == "false":
```
And I still don't get this idea. Why do we check again the file URL?

Reply:
In the case the user did copy from screenshot(or from image in apps), the clipboard will actually store image data but not file path, so if we don't check this, the addon will still paste as if user copied a single image from Finder, which doesn't exist.

Also, applescript is weird, so we actually needed to use `contains` instead of `contain`, like so:
```
(((clipboard info) as string) contains "«class furl»") as string
```

Reply:
_applescript_ seems to be an interesting scripting language though :laughing: 

> In the case the user did copy from screenshot(or from image in apps), the clipboard will actually store image data but not file path, so if we don't check this, the addon will still paste as if user copied a single image from Finder, which doesn't exist.

But did we check it here:
https://github.com/Yeetus3141/ImagePaste/blob/0579010a527980580b62b7560f9c2647bfc0d4cd/imagepaste/clipboard/darwin/darwin.py#L37-L40

And then did we check if clipboard contains real image data:
https://github.com/Yeetus3141/ImagePaste/blob/0579010a527980580b62b7560f9c2647bfc0d4cd/imagepaste/clipboard/darwin/darwin.py#L45-L48

Reply:
Oh yes we did 😅, probably I didn't check careful enough because pasteboard doesn't return actual image data so I resorted to applescript.

So in that case I think we could just check if `urls is None` then if that's the case invoke the applescript to paste single image

Reply:
I still get confused because `urls is None` is not guaranteed for a real image data, clipboard could contain text data, sound data,... And what do we receive from `contents` if clipboard contains those types?

BTW,  do you mean anything to remove these lines in 6baefa96bfb734ea7ca819aa29e53c5747eecf35:
```python
            if process.stderr:
                report = Report(6, f"Saved 1 image: {image} (WARN: {process.stderr})")
                return cls(report, [image])
```

Reply:
Oh no, I didn't mean to for https://github.com/Yeetus3141/ImagePaste/commit/6baefa96bfb734ea7ca819aa29e53c5747eecf35 , will amend

Maybe I should clarify this, `get_contents` will always return a string from clipboard, according the pasteboard documentation: https://github.com/tobywf/pasteboard

Reply:
I've just looked over and I wonder if it's OK if we do this:
```diff
 class DarwinClipboard(Clipboard): 
     @classmethod
     def push(cls, save_directory: str) -> DarwinClipboard:
         from os.path import join
+        from os.path import isfile
         from .pasteboard._native import Pasteboard
+        from .pasteboard._native import PasteboardType
 
-        # Use Pasteboard to get file URLs from the clipboard
         pasteboard = Pasteboard()
+
+        # Use Pasteboard to get file URLs from the clipboard
         urls = pasteboard.get_file_urls()
         if urls is not None:
             filepaths = list(urls)
             images = [Image(filepath) for filepath in filepaths]
             return cls(Report(6, f"Pasted {len(images)} image files: {images}"), images)
 
-        # If no images are found, return a report with no images
-        contents = pasteboard.get_contents()
-        if contents == "":
-            return cls(Report(2))
-
-        # Check if clipboard doesn't contain any filepaths
-        # (e.g. if the clipboard contains just a single image)
-        commands = [
-            '((clipboard info) as string does not contain "«class furl»") as string'
-        ]
-        process = Process(cls.get_osascript_args(commands))
-        if process.stdout == "true":
+        # Save an image if it is in the clipboard
+        contents = pasteboard.get_contents(type=PasteboardType.PNG)
+        if contents is not None:
             filename = cls.get_timestamp_filename()
             filepath = join(save_directory, filename)
             image = Image(filepath, filename)
             commands = [
                 "set pastedImage to "
                 f'(open for access POSIX file "{filepath}" with write permission)',
                 "try",
                 "    write (the clipboard as «class PNGf») to pastedImage",
                 "end try",
                 "close access pastedImage",
             ]
             process = Process.execute(cls.get_osascript_args(commands))
-            if process.stderr:
+            if not isfile(filepath):
                 return cls(Report(3, f"Cannot save image: {image} ({process.stderr})"))
+            if process.stderr:
+                report = Report(6, f"Saved 1 image: {image} (WARN: {process.stderr})")
+                return cls(report, [image])
             return cls(Report(6, f"Saved and pasted 1 image: {image}"), [image])
-        return cls(Report(3))
+        return cls(Report(2))
```

Reply:
I will have to test this out,  mainly I am also not sure if other image types will get passed, also will have to check since we imported the `pasteboard` module directly we might need to use _native.PNG, based on what I have seen here: https://github.com/tobywf/pasteboard/blob/master/tests.py

Reply:
I agree with you that it only writes to PNG with any types of image, but because PNG is a rich format, we might partly meet our target.

I've used `xclip` to handle clipboard on _X11_, it only targets to `image/png` and it can copy-paste all image that browser can display on [this website](https://wiesmann.codiferes.net/share/bitmaps/). Does `pateboard` act the same on your computer? If it does, we can keep the type checking method as it provides safer operations, avoiding wrong error catching.

Reply:
Just tested it out, targeting PNG only works for screenshots/copying image from other apps but however setting to TIFF also copy paste all images(as supported by the browser) on that website correctly, like so:

```python
from .pasteboard._native import TIFF
...
contents = Pasteboard().get_contents(type=_native.TIFF)
```

Reply:
That's great, I guess we're coming to an end of this fix. With the commit c8d41ce9230234ea1ccdd2db8fbb0cb4ff37f9e0 I've just pushed, do you think it's ready to merge? Anything you feel we should have another thought?

Reply:
Yep I think that's all good for now. Thanks for your work!

Reply:
No, you did the job. Thank you for taking time to support us. Let me know if you have any idea, I really appreciate it.